Add GlobalValueNumbering pass

hkapp opened the pull request:
Add the whole implementation for GVN, comprised of the definition of idempotent (might be a problem, please look at the comment), deep equality of variables, and the pass itself.

The Debug part that are useful should be put somewhere else, hints on "where" would be appreciated.

densh commented:
You forgot to mention that this fixes #32.

densh commented on tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala:
```diff
@@ -120,4 +393,91 @@ object GlobalValueNumbering {
 
   }
 
+  object Debug {
```
Those should go to `Shows`.

hkapp commented:
I can't put the Debug methods in nir.Shows, as nir does not depend on tools (and it really shouldn't). Any ideas ?

densh commented:
@hkapp Lets put them in `compiler.analysis.Shows` then.

hkapp commented:
@densh Ok, fixed it. What would you say about the comment I added in `isIdempotent` ? I really think that this invalidates the definition of idempotency for `call`, and that we won't be able to reuse these values until there is static analysis available for this specific purpose.

densh commented on tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala:
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
+  }
+
+  private def pred(block: Block) = block.pred.map(_.from)
+  private def succ(block: Block) = block.succ.map(_.to)
```
Lets rename `block.pred` into `block.inEdges` and `block.succ` into `block.outEdges` and then put `def pred` and `def succ` as helper methods directly inside `ControlFlow.Block`.

densh commented on tools/src/main/scala/scala/scalanative/compiler/analysis/Shows.scala:
```diff
@@ -0,0 +1,52 @@
+  def showCFG(cfg: ControlFlow.Graph): String = {
+    cfg.all.map { block =>
+      val succStr =
+        block.succ.map(_.to).map(blockToString).mkString("(", ",", ")")
```
`block.succ.map(_.to)` can be simplified to `block.pred` after refactoring I've suggested earlier.

densh commented on tools/src/main/scala/scala/scalanative/compiler/analysis/Shows.scala:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  def showDominatorTree(domination: Map[Block, Set[Block]]): String = {
+    domination.toList
```
`toSeq`

densh commented on tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala:
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
+   * a.b = 1
+   * val d = a.toString
+   * --> toString is pure, but the values of c and d are different !
+   * */
```
`toString` is not pure, `load`-s on external memory are side-effecting by definition.

densh commented on tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala:
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
+      case Call(_, Val.Global(Ref(node), _), _) =>
+        node.attrs.isPure
+
+      case Call(_, _, _) => false
```
`case _: Call => false`

densh commented on tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala:
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
+
+      case Call(_, _, _) => false
+
+      case _: Closure => ???
```
Lets assume that this one is not idempotent for now.

densh commented on tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala:
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
+            eqType(tyA, tyB) && eqVal(ptrA, ptrB) && eqVals(argsA, argsB)
+
+          case (Load(_, _), Load(_, _)) =>
+            false // not idempotent
```
It seems that this and all the other not-idempotent cases are never going to happen due to the check above, therefore we should drop this code altogether.

hkapp commented:
Looking at the code in ControlFlow.Block, there are still a few variables named `node(s)` instead of `block`. I also saw that most of them have been changed, but in `Graph.apply` they are both used to denote different things. Any semantic meaning to this ?

densh commented:
We used to have `ControFlow.Node`, that has been subsumed by `ControlFlow.Block`. Node is a block, and visa versa.

densh commented:
LGTM, great job @hkapp !
